bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,278 @@
+                applyExpressionParts();
+            }
+            continueSearch = false;
+        }
+
+        for (Date aDate : getCandidates()) {
+            logger.trace("Final candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+        }
+    }
+
+    abstract protected void validateExpression() throws IllegalArgumentException;
+
+    protected void applyExpressionParts() {
+        Collections.sort(getExpressionParts());
+        for (ExpressionPart part : getExpressionParts()) {
+            logger.trace("Expanding {} from {} candidates", part.getClass().getSimpleName(), getCandidates().size());
+            setCandidates(part.apply(startDate, getCandidates()));
+            logger.trace("Expanded to {} candidates", getCandidates().size());
+            for (Date aDate : getCandidates()) {
+                logger.trace("Candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+            }
+            prune();
+        }
+    }
+
+    protected void prune() {
+        Collections.sort(getCandidates());
+
+        ArrayList<Date> beforeDates = new ArrayList<Date>();
+
+        for (Date candidate : getCandidates()) {
+            if (candidate.before(startDate)) {
+                beforeDates.add(candidate);
+            }
+        }
+
+        getCandidates().removeAll(beforeDates);
+
+        if (getCandidates().size() > maximumCandidates) {
+            logger.trace("Pruning {} candidates to {}", getCandidates().size(), maximumCandidates);
+            int size = getCandidates().size();
+            for (int i = maximumCandidates; i < size; i++) {
+                getCandidates().remove(getCandidates().size() - 1);
+            }
+        }
+    }
+
+    @Override
+    public Date getTimeAfter(Date afterTime) {
+        if (getCandidates().isEmpty()) {
+            try {
+                setStartDate(afterTime);
+                parseExpression(expression);
+            } catch (ParseException e) {
+                logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
+            }
+        }
+
+        if (!getCandidates().isEmpty()) {
+
+            Collections.sort(getCandidates());
+
+            for (Date candidate : getCandidates()) {
+                if (candidate.after(afterTime)) {
+                    return candidate;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public Date getFinalFireTime() {
+        if (getCandidates().isEmpty()) {
+            try {
+                parseExpression(getExpression());
+            } catch (ParseException e) {
+                logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
+            }
+        }
+
+        if (getCandidates().isEmpty()) {
+            return null;
+        }
+
+        return getCandidates().get(getCandidates().size() - 1);
+    }
+
+    abstract protected E parseToken(String token, int position) throws ParseException;
```

Comment:
Could you add a JavaDoc to this function as it has to be implemented by the sub class correctly.